New article parser

tgalery opened the pull request:
Hi @diegoceccarelli, this is a PR that abstracts away some of the work done at Idio and by Gsoc on a new stats extractor for spotlight. I thought it would be nice to try to upstream some of the changes, so we can keep things centralised with you.

A couple of things to note. We changed `description` to a more natural form of `anchor`, but if that breaks stuff (e.g. dexter), I'd be happy to change it back to the original.

We also made some of the article atributes transient, but we could revert that if some of them are necessary (e.g. I think this does some template unpacking, so they are not saved at the end result).

I'm gonna make questions myself to people involved in this like @dav009 and @naveenmadhire , so please fire any questions away (I wouldn't merge this until they are properly discussed).

tgalery commented on pom.xml:
```diff
@@ -121,6 +121,14 @@
 			<url>https://oss.sonatype.org/content/repositories/snapshots</url>
 		</repository>
 		<repository>
+			<id>jitpack.io</id>
```
Why do we need jitpack again ?

dav009 replied:
so that you can use jsonwikipedia in another pom/sbt project?
They act as some sort of mvn repo in which you can specify a url to a github repo and it will build it and give maven the jar of it.

dav009 replied:
Probably can be removed now. Before we had our own repo with bliki which I guess it is the reason why jitpack was introduced

tgalery replied:
Gonna sack it off then.

tgalery commented on src/main/java/it/cnr/isti/hpc/wikipedia/article/Article.java:
```diff
@@ -53,18 +53,19 @@
-	private List<String> templatesSchema;
-	private List<String> highlights;
+	private transient List<String> sections;
+	private transient List<String> paragraphs;
```
I think paragraphs might be needed, since we might want to create 'fake links' on text to get more reliable stats for entity linking, what do you reckon @naveenmadhire @dav009 ?

naveenmadhire replied:
The new variable paragraphWithLinks will have Paragragh text in it along with links associated with that paragraph. Hence having paragraph text in a different variable would be redundant. Does that make sense?

diegoceccarelli replied:
What about keeping the paragraphs, but add in the links an int with the position in the text?

dav009 replied:
yes, that make sense @naveenmadhire

tgalery commented on src/main/java/it/cnr/isti/hpc/wikipedia/article/Link.java:
```diff
@@ -50,19 +53,18 @@ public void setId(String id) {
-            this.description = description;
-        }
+        if (anchor == null || anchor.isEmpty())
+           this.anchor = this.id.replace("_", " ");
```
Is this the mechanism for handling empty links ? If so, I wonder whether this would not generate false links for a system that presuppose that the entities linked would appear in wikipedia.

dav009 replied:
No, this is just handling annotations without anchors like [[Barack_Obama]]
For some reason whatever component generates them, was sending also `nulls` so when you built the jsonpedia it would output some exceptions.
Thus the extra check was added

tgalery replied:
cool

tgalery commented on src/main/java/it/cnr/isti/hpc/wikipedia/parser/ArticleParser.java:
```diff
@@ -15,30 +15,23 @@
 package it.cnr.isti.hpc.wikipedia.parser;
 
-import it.cnr.isti.hpc.wikipedia.article.Article;
+import de.tudarmstadt.ukp.wikipedia.parser.*;
```
I actually prefer the old imports, so I might change these back.

dav009 replied:
yup, it is a good call

tgalery commented on src/test/java/it/cnr/isti/hpc/wikipedia/article/en/ArticleTest.java:
```diff
@@ -44,10 +53,28 @@ public void testParsing() throws IOException {
-		assertEquals(5, a.getCategories().size());
-		assertEquals(7,a.getSections().size());
-		assertEquals(74,a.getLinks().size());
+//		assertTrue("Wrong parsed text",a.getCleanText().trim().startsWith("Albedo (), or reflection coefficient, is the diffuse reflectivity or reflecting power of a surface."));
```
Don't understand why this test is commented out. Should it be put back in ?

naveenmadhire replied:
Not sure about this, @dav009 may know.

tgalery commented on src/test/java/it/cnr/isti/hpc/wikipedia/article/en/ArticleTest.java:
```diff
@@ -64,15 +91,15 @@ public void testMercedes() throws IOException {
-		
-	}
-	
+//@Test
```
Same thing here, not sure why test has been commented out.

tgalery commented on src/test/java/it/cnr/isti/hpc/wikipedia/article/en/ArticleTest.java:
```diff
@@ -88,24 +115,113 @@ public void testNotRedirect() throws IOException {
-        for (Link link:a.getLinks()){
-            assertFalse(link.getDescription().isEmpty());
+        for (Link link : a.getLinks()) {
+            assert (link.getAnchor()!="");
```
more robust test ? Why changing from `isEmpty` to `!= ''` ?

diegoceccarelli replied:
use isEmpty please, !="" could be wrong

tgalery replied:
will do

tgalery commented on src/test/java/it/cnr/isti/hpc/wikipedia/article/it/ArticleTest.java:
```diff
@@ -61,35 +60,35 @@ public void categories() throws IOException {
-		
-
-	}
+	//@Test
```
Why commented out test ?

dav009 replied:
yeah, sorry. 
There were some changes in our branch that would have us to update the fixtures for italian.
This was commented cause it was not our priority at the given moment

diegoceccarelli commented on src/test/java/it/cnr/isti/hpc/wikipedia/article/en/ArticleTest.java:
```diff
@@ -88,24 +115,113 @@ public void testNotRedirect() throws IOException {
+
+        for (ParagraphWithLinks p : a.getParagraphsWithLinks()) {
+            for(Link link:p.getLinks()){
+                assert (!link.getId().equals(""));
```
use .isEmpty() please

diegoceccarelli commented:
Hi Guys, thanks for the PR. I definitely like the idea of having links per paragraph. Some random thoughs: 
  - I'm not sure that I like the transients and that you removed some attributes from the Article class. What about having a 'configuration' object that tells to the parser which attributes we want to serialize? (I Have to take a look but probably this is supported in the json library) what do you think? 
 - Ok for changing the name from description -> anchor  I like it, it is more clear. Let's just make a new version so that I can compile the current dexter version with the previous version and fix the new version of dexter for using the new schema. 
 - @dav009 et @naveenmadhire @keynmol.  I think we should meet one of these days :) 
 - I was thinking to move from GSON to another json library, like json noggit. It could produce significant performance speed-up.

dav009 commented:
@diegoceccarelli @tgalery sorry, our idio branch is somewhat "stable" at the moment.
We introduced some changes which cleans lots of noise, and which speeds up the task by parallelizing it.
However on the road we did remove some of the tests for italian.

This rebase seem to miss some of those changes, which can be get by rebasing against idio's fork

tgalery commented:
Hey guys thanks for the comments, so here is what I plan to do:

- @diego: I like the idea of supporting what to serialize in a configuration object too, I'm not sure whether that is linked with the possible change of the library, but for the moment, I will just remove the transients and leave that as part of another PR.

-  @dav009 @naveenmadhire: given @diegoceccarelli 's comments about the ParagraphWithLinks class, I wonder whether we should just keep the original Paragraph and have a method that puts the links in the paragraphs dynamically. It could be even something that returns a ParagraphsWithLinks, but that would be outside the schema? I'm also saying that because this method could be used in the generation of 'new links' for matches of the same surface form in the same article.

- @dav009 The initial rebase that generated this PR just selected the commits of the Idio repo touching article and link classes. I think that another rebase to incorporate your commits that fix the tests would be quite painful, so I might take a look at your test fixing commits and add them here manually (since I am changing some stuff). That being said, your work on the parallel stuff would be a great contribution, I'm thinking of doing the same type or rebase and sending a PR over, but if any of you is faster than me, I won't be offended ;-).

I'm gonna be in London in Jan, let's go for :beers: !

tgalery commented:
Hi @naveenmadhire  @dav009 and @diegoceccarelli ,  here is an update:

- I have removed the 'transient' annotations, so we have the output serialized as is now. Once we have a config file we could change how the serialization is returned.
- I sacked of paragraphs and stuck with paragraphsWithLinks. The reason being that start and end positions are relative to paragraphs, not full text. By having each paragraph with a text and a links attribute we can easily test and determine whether these get captured in the right way.
- replaced some methods by 'isEmpty' as suggested by diego and naveen
- Put some old commented out code back

However there are two old test cases testDisambiguation() and testInfobox() which have been commented out but fail when I uncomment them. At least the latter is still commented out in the Idio codebase, so I'd suggest to take a look at those with more care before we merge this.

diegoceccarelli commented:
:+1: thanks @tgalery, I'll take a look during the week

diegoceccarelli commented:
Quick update, @tgalery and @dav009 :
   1. I checked out your branch and I'm reviewing it, I'll soon push a pull request on you branch, with my comments/changes
   2. I'm thinking about the paragraphes with links. I like to have the position of the links, but the change will produce a not retrocompatible schema and that could be annoying for someone. I'm not against changing the schema but.. could be possible to avoid? what if in the current links we put the absolute/relative position of each  link (start/end) and the paragraph number? 
  3 what if a link is in a list/table/template?  
  4 @dav009  while I was working on the new version of Dexter, I noticed that I had some anchor text in a record that I was not able to match in a paragraph, or inside lists/tables.. (I'm using the 'simple' English dump) did you experienced similar cases? if yes, how they are managed?

tgalery commented:
Hi @diegoceccarelli sorry for taking so long to reply to you. Here are some thoughts:
- I think I'm cool with adding a `paragraphIndex` to the link class so we can do the substitutions (start and end position are already there).
- I think some of the commits that  @dav009 unpack the textual content of lists and tables as paragraphs (after the main paragraphs are extracted). But maybe he can correct me.

I'm happy to make those changes if you want (or else find people who might contribute to the project, starting with this PR specifically).

Let me know....

diegoceccarelli commented:
Dear @tgalery and @dav009,
If we can just /enrich/ the current schema it would be cool so it would be also retrocompatible. I was thinking that each link could have a 'type' (Paragraph, Table, List, Infobox etc). @dav009  I would like to keep Paragraphs separated from Table, List .. and then maybe have a method in Article that allows you to iterate on all the text if needed. @tgalery In this period I'm working on the new version of dexter (..and I think you could be really interested once that is ready ;)) but if you, or someone else can contribute I'll be happy to review and merge. 
@dav009 did you experience the issue with the links that are not in the paragraph/tables/list etc?

tgalery commented:
Cheers @diegoceccarelli, I'm just wondering what would be the best approach. Maybe it would be best to create a branch from scratch and do the following:

- introduce (i) start, (ii) end, (iii) paragraphIndex, and paragraphType to Link.java
- introduce paragrapthType to paragraph (type would be 'body', 'table' and 'list')
- add functions that grab paragraphs from tables and lists and puts them in the list of paragraphs 

I think @rod3go might be interested in working with this.

diegoceccarelli commented:
What I was thinking was to have: 
   * Paragraphs: containing a list of paragraphes how as the moment
   * Tables & List as a the moment
   * Then in each link we could have startPosition, endPosition and type=body,table,list and extra attributes based on the type: 
         - if type == body: paragraphId
         - if type == list: listId (there could be several lists) + itemId
         - if type == table: tableId (there could be several table) + columnId+rowId

What do you think? 
For the development, I agree on a new branch, I recently discovered that on github we can 
have 'enhancement issues' that works as jira items, so I can create a story and we can create 
a feature-branch and commits referring to the story. How does it sound? 
Cheers,
Diego

tgalery commented:
Ok, so @rod3go is starting a new branch on this, for starters, we are adding paragraphIds and a type, it probably would be interesting to add 'category' and 'image' as link types for links found in these contexts. We then can have a chat about how to best represent listIds and table ids, migh feeling is that @dav009 just grabbed the text and removed non-textual structure cause it could interfere with the generation of distributional models (word2vec) (a solution to that would be to include listParagraphs or TableParagraphs as part of an article). Let me know your thoughts.

diegoceccarelli commented:
@tgalery I like the idea of having category and image link types, would be cool to be able to access the link to the images for each article. @rod3go please open PR as soon as you have something, also if it is preliminary :)  I create an item in #16 to keep track of the progress, please add #16 in the commit messages :)

rod3go commented:
@diegoceccarelli and @tgalery, regarding tables, how should I treat links found at the title of a table, assuming that rowId and columnId both start at 0?
* Ignore?
* Create a new Type?
* Something else?

tgalery commented:
@diegoceccarelli @rod3go I think we should move this discussion to #17, so I will comment there instead!

tgalery commented:
closing due to #27
